### PR TITLE
Implement systems for 1828

### PR DESCRIPTION
**Pull request by crericha**
Main things added:
- System merger process
- Support to discard trains from shells
- Support to buy trains in each shell
- Include system treasury share in the dividend calculations
- Support up to 8 train routes

### CODE REVIEW

File: assets/app/lib/settings.rb [hunk 1]
```diff
@@ -7,7 +7,7 @@ module Settings
     DARK = `window.matchMedia('(prefers-color-scheme: dark)').matches`.freeze
     # http://mkweb.bcgsc.ca/colorblind/ 12 color palette
     # change SETTINGS = (n.times ... in models/user.rb accordingly
-    ROUTE_COLORS = %i[#A40122 #008DF9 #00FCCF #FF5AAF #8400CD #FF6E3A].freeze
+    ROUTE_COLORS = %i[#A40122 #008DF9 #00FCCF #FF5AAF #8400CD #FF6E3A #009F81 #FFC33B].freeze
```
Comment (crericha) [inline assets/app/lib/settings.rb:10]: Systems can run up to 8 trains. Add more route colors from the colorblind palate.

File: assets/app/view/game/buy_sell_shares.rb [hunk 2]
```diff
@@ -25,6 +25,7 @@ def render
         children = []
 
         children.concat(render_buy_shares)
+        children.concat(render_merge)
```
Comment (crericha) [inline assets/app/view/game/buy_sell_shares.rb:28]: Trigger to transition within stock round to the merge interface used by 1817.

File: assets/app/view/game/buy_trains.rb [hunk 3]
```diff
@@ -10,12 +10,15 @@ class BuyTrains < Snabberb::Component
...
       needs :show_other_players, default: nil, store: true
       needs :selected_corporation, default: nil, store: true
+      needs :active_shell, default: nil, store: true
 
       def render_president_contributions
         player = @corporation.owner
 
         children = []
 
+        verb = @must_buy_train ? 'must' : 'may'
```
Comment (crericha) [inline assets/app/view/game/buy_trains.rb:20]: President can contribute funds to buy a train in the shell if the shell is empty

File: assets/app/view/game/buy_trains.rb [hunk 4]
```diff
@@ -288,6 +303,30 @@ def remaining_trains
           *rows,
         ])
       end
+
```
Comment (crericha) [inline assets/app/view/game/buy_trains.rb:306]: Add buttons above the train buy elements to select which shell the player is buying for and store within the page. The train buying view will use this information to decide whether or not a train is buyable in that shell and if the president can contribute funds.

File: assets/app/view/game/discard_trains.rb [hunk 5]
```diff
@@ -15,8 +15,9 @@ def render
             verticalAlign: 'top',
           },
         }
-        overflow = @game.active_step.crowded_corps.map do |corporation|
-          trains = corporation.trains.map do |train|
+        step = @game.active_step
+        overflow = step.crowded_corps.map do |corporation|
+          trains = step.trains(corporation).map do |train|
```
Comment (crericha) [inline assets/app/view/game/discard_trains.rb:20]: Get the trains from the step, so shells can be handled without the view needing to know about them.

File: assets/app/view/game/round/operating.rb [hunk 6]
```diff
@@ -45,6 +45,8 @@ def render
             left << h(CorporateSellShares)
           elsif @current_actions.include?('choose')
             left << h(Choose)
+          elsif @current_actions.include?('swap_train')
+            left << h(SwapTrain)
```
Comment (crericha) [inline assets/app/view/game/round/operating.rb:49]: Support for new step that allows swapping of trains between shells of a system.

File: assets/app/view/game/sell_shares.rb [hunk 7]
```diff
@@ -12,6 +12,9 @@ class SellShares < Snabberb::Component
       needs :action, default: Engine::Action::SellShares
 
       def render
+        step = @game.round.active_step
+        return unless step.current_actions.include?('sell_shares')
```
Comment (crericha) [inline assets/app/view/game/sell_shares.rb:16]: Don't render unless the action is included. I'm not sure why this guard didn't already exist. Other stock round action rendering logic has it.

File: assets/app/view/game/sell_shares.rb [hunk 8]
```diff
@@ -36,7 +39,6 @@ def render
           )
         end
 
-        step = @game.round.active_step
```
Comment (crericha) [inline assets/app/view/game/sell_shares.rb:39]: Step is needed. Will add back in.

File: assets/app/view/game_page.rb [hunk 9]
```diff
@@ -255,7 +255,7 @@ def render_action
 
       case @round
       when Engine::Round::Stock
-        if (%w[place_token lay_tile] & current_actions).any?
+        if (%w[place_token lay_tile remove_token] & current_actions).any?
```
Comment (crericha) [inline assets/app/view/game_page.rb:258]: Add support for remove_token action during the stock round

File: lib/engine/action/start_merge.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+      end
+
+      def args_to_h
+        {
+          'corporation' => @corporation.id,
+        }
+      end
+    end
+  end
+end
```
Comment (crericha) [inline lib/engine/action/start_merge.rb:28]: Action associated with selecting the first corporation for a merger (done during the stock round). After the corporation is selected, the game transitions to the 1817-like merge view to select the second corporation for the merger.

File: lib/engine/corporation.rb [hunk 11]
```diff
@@ -25,13 +25,14 @@ class Corporation
     include Transfer
 
     attr_accessor :ipoed, :par_via_exchange, :max_ownership_percent, :float_percent, :capitalization, :max_share_price
-    attr_reader :companies, :min_price, :name, :full_name, :fraction_shares, :type
+    attr_reader :companies, :min_price, :name, :full_name, :fraction_shares, :type, :id
```
Comment (crericha) [inline lib/engine/corporation.rb:28]: While in the middle of the merger, the system and the original corporation have the same name. Allow id to be different than name.

File: lib/engine/game/base.rb [hunk 12]
```diff
@@ -1388,15 +1388,15 @@ def corporation_opts
...
-        end
+        self.class::CORPORATIONS.map { |corporation| init_corporation(stock_market, corporation) }
+      end
+
+      def init_corporation(stock_market, corporation)
+        Corporation.new(
+          min_price: stock_market.par_prices.map(&:price).min,
+          capitalization: self.class::CAPITALIZATION,
+          **corporation.merge(corporation_opts),
+        )
```
Comment (crericha) [inline lib/engine/game/base.rb:1399]: Allow for initialization of one corp

File: lib/engine/player.rb [hunk 13]
```diff
@@ -36,7 +36,9 @@ def player
     end
 
     def ==(other)
-      @name == other&.name
+      return false if !other || !other.player?
+
+      @name == other.name
```
Comment (crericha) [inline lib/engine/player.rb:41]: Add safe guard, so player can't be equal to something other than a player.

File: lib/engine/step/g_1817/acquire.rb [hunk 14]
```diff
@@ -51,6 +51,7 @@ def can_take_loan?(entity)
         end
 
         def can_payoff?(entity)
+          @unpaid_loans ||= 0
```
Comment (crericha) [inline lib/engine/step/g_1817/acquire.rb:54]: Will remove. This was used for debugging.

File: lib/engine/step/g_1828/buy_sell_par_shares.rb [hunk 15]
```diff
@@ -6,13 +6,44 @@ module Engine
...
+            actions << 'pass' if actions.any? && !actions.include?('pass')
+          end
+
+          actions
+        end
+
+        def process_start_merge(action)
+          @game.game_error('No eligible corporation to merge with') unless can_merge?(action.entity, action.corporation)
+          @round.merge_initiator = action.corporation
+          @round.acting_player = action.entity
```
Comment (crericha) [inline lib/engine/step/g_1828/buy_sell_par_shares.rb:26]: round variables used to trigger merger step

File: models/user.rb [hunk 16]
```diff
@@ -10,7 +10,7 @@ class User < Base
 
   RESET_WINDOW = 60 * 15 # 15 minutes
 
-  SETTINGS = (6.times.flat_map do |index|
+  SETTINGS = (8.times.flat_map do |index|
```
Comment (crericha) [inline models/user.rb:13]: Other part of supporting up to 8 routes.

File: lib/engine/step/g_1828/merger.rb [hunk 17]
```diff
@@ -0,0 +1,448 @@
+# frozen_string_literal: true
```
Comment (crericha) [inline lib/engine/step/g_1828/merger.rb:1]: Contains the merge state machine that allows the player to select the merge target, creates the merged corporation, and handles share exchanges, and complete the merger. The exchange logic may require decisions by the players (who to trade a share with, which share to keep if an odd percent owned) and is designed to halt when a player choice is required and then pick back up after it is made.

Waiting on a few rule updates from JC and the multi-turn undo to completely finish and go to pre-alpha.

File: assets/app/view/game/buy_trains.rb [hunk 18]
```diff
@@ -92,9 +97,16 @@ def render
           },
         }
 
-        if (step.can_buy_train?(@corporation) && step.room?(@corporation)) || must_buy_train
-          children << h(:div, "#{@corporation.name} must buy an available train") if must_buy_train
-          if should_buy_train == :liquidation
+        if @corporation.system?
+          children << render_shells
+        else
+          store(:active_shell, nil)
```
Comment (tobymao) [inline assets/app/view/game/buy_trains.rb:103]: do you want this to cause a rerender?

File: lib/engine/action/buy_train.rb [hunk 19]
```diff
@@ -30,8 +33,21 @@ def args_to_h
           'price' => @price,
           'variant' => @variant,
           'exchange' => @exchange&.id,
+          'shell' => @shell ? @shell.name : nil,
```
Comment (tobymao) [inline lib/engine/action/buy_train.rb:36]: @shell&.name

File: lib/engine/action/start_merge.rb (hunk 10 above)
Comment (tobymao) [reply lib/engine/action/start_merge.rb:28]: do you need a new action? could you use choose?

File: lib/engine/corporation.rb (hunk 11 above)
Comment (tobymao) [reply lib/engine/corporation.rb:28]: what happens if you change corporation id to be sym?

File: lib/engine/g_1828/shell.rb [hunk 20]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module Engine
+  module G1828
+    class Shell
+      attr_accessor :name, :trains, :system
```
Comment (tobymao) [inline lib/engine/g_1828/shell.rb:6]: trains should be a reader right? do you actually need to make a new array?

Comment (tobymao) [reply lib/engine/g_1828/shell.rb:6]: what exactly is a shell?

File: models/user.rb (hunk 16 above)
Comment (tobymao) [reply models/user.rb:13]: hm this shouldn't be hardcoded, can you pull the size of the colors here?

File: lib/engine/corporation.rb (hunk 11 above)
Comment (crericha) [reply lib/engine/corporation.rb:28]: Like name.to_sym? That should work fine. I can test it.

File: lib/engine/action/start_merge.rb (hunk 10 above)
Comment (crericha) [reply lib/engine/action/start_merge.rb:28]: Makes sense. I'll use choose.

File: models/user.rb (hunk 16 above)
Comment (crericha) [reply models/user.rb:13]: I agree. yzemaze hardcoded it for some reason. I can ask him. Are you okay with models dir files referencing assets dir files?

File: assets/app/view/game/buy_trains.rb (hunk 18 above)
Comment (crericha) [reply assets/app/view/game/buy_trains.rb:103]: To be honest, I don't understand what's going on under-the-hood with settings. How long is the setting around for? If an action occurs, is it wiped? That's what I'm trying to do here.

File: lib/engine/action/buy_train.rb (hunk 19 above)
Comment (crericha) [reply lib/engine/action/buy_train.rb:36]: Agree

File: lib/engine/g_1828/shell.rb (hunk 20 above)
Comment (crericha) [reply lib/engine/g_1828/shell.rb:6]: Yes, trains should be a reader

File: assets/app/view/game/buy_trains.rb (hunk 18 above)
Comment (tobymao) [reply assets/app/view/game/buy_trains.rb:103]: add skip: true

File: assets/app/view/game/swap_train.rb [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+          style: {
+            display: 'grid',
+            grid: 'auto / repeat(2, max-content)',
+            gap: '0.5rem',
+          },
+        }
+
+        children = []
+
+        @corporation.shells.each do |shell|
```
Comment (tobymao) [inline assets/app/view/game/swap_train.rb:29]: children = @corporations.shelss.flat_map

File: lib/engine/corporation.rb (hunk 11 above)
Comment (tobymao) [reply lib/engine/corporation.rb:28]: asking about using sym as the id. 

can you just make a fake name instead? which is like name + ' system'

File: lib/engine/game/g_1828.rb [hunk 22]
```diff
@@ -221,6 +230,131 @@ def upgrades_to?(from, to, special = false)
...
+
+        @round.players_sold[entity][corporation]
+      end
+
+      def create_system(corporations)
+        system_data = CORPORATIONS.find { |c| c['sym'] == corporations.first.id }.dup
+        system_data['sym'] = corporations.map(&:name).join('-')
+        system_data['tokens'] = []
+        system = init_corporation(@stock_market, system_data)
+        system.extend(Engine::G1828::System)
```
Comment (tobymao) [inline lib/engine/game/g_1828.rb:265]: why do you need to extend it like this?, why can't you just do System.new?

File: lib/engine/game/g_1828.rb [hunk 23]
```diff
@@ -221,6 +230,131 @@ def upgrades_to?(from, to, special = false)
...
+          end
+          system.shells << shell
+
+          corporation.companies.each do |company|
+            company.owner = system
+            system.companies << company
+          end
+          corporation.companies.clear
+
+          corporation.all_abilities.each do |ability|
```
Comment (tobymao) [inline lib/engine/game/g_1828.rb:298]: no issues with removing or adding abilities while iterating?

File: lib/engine/game/g_1828.rb [hunk 24]
```diff
@@ -221,6 +230,131 @@ def upgrades_to?(from, to, special = false)
...
+        end
+
+        @stock_market.set_par(system, system_market_price(corporations))
+        system.ipoed = true
+
+        system
+      end
+
+      def create_system_tokens(system, corporations)
+        used, unused = corporations.map(&:tokens).flatten.partition(&:used)
```
Comment (tobymao) [inline lib/engine/game/g_1828.rb:311]: flat_map

File: models/user.rb (hunk 16 above)
Comment (tobymao) [reply models/user.rb:13]: maybe

File: lib/engine/corporation.rb (hunk 11 above)
Comment (crericha) [reply lib/engine/corporation.rb:28]: That's what I'm doing. This is to get around the fact that sym is also @name, which is displayed on charters. I split them into different variables (line 35) and then overwrite the id to something unique for the system.

File: lib/engine/game/g_1828.rb (hunk 23 above)
Comment (crericha) [reply lib/engine/game/g_1828.rb:298]: I'll look into it. Probably needs a dup. I don't think there were any abilities in my test case.

File: lib/engine/game/g_1828.rb (hunk 24 above)
Comment (crericha) [reply lib/engine/game/g_1828.rb:311]: One day I'll remember that method exists. ;-)

File: lib/engine/game/g_1828.rb (hunk 22 above)
Comment (crericha) [reply lib/engine/game/g_1828.rb:265]: I felt like putting the the corporation initialization logic into the 1828 game class was a fragile design. If anyone ever changed it, 1828 would likely break. This way, I'm decorating my additions and leaving everything else alone. I actually thought this was kind of elegant. What do you think?

Comment (tobymao) [reply lib/engine/game/g_1828.rb:265]: what do you mean? why can't there just be an 1828 specific corporation?

File: assets/app/view/game/swap_train.rb (hunk 21 above)
Comment (crericha) [reply assets/app/view/game/swap_train.rb:29]: Fixed.

File: lib/engine/game/g_1828.rb (hunk 22 above)
Comment (crericha) [reply lib/engine/game/g_1828.rb:265]: Are you suggesting inherit from init_corporations and call System.new from there? That would be fine. What I don't want to do is duplicate the init logic.

File: lib/engine/game/g_1828.rb (hunk 23 above)
Comment (crericha) [reply lib/engine/game/g_1828.rb:298]: Fixed.

File: lib/engine/game/g_1828.rb (hunk 22 above)
Comment (tobymao) [reply lib/engine/game/g_1828.rb:265]: what do you mean inherit from init_corporations? how do you inherit a method?

Comment (crericha) [reply lib/engine/game/g_1828.rb:265]: Bad language -- it was late last night. I meant override.

Comment (crericha) [reply lib/engine/game/g_1828.rb]: Refactored.

File: lib/engine/step/g_1828/merger.rb (hunk 17 above)
Comment (crericha) [reply lib/engine/step/g_1828/merger.rb:1]: Implemented based on a conversation with JC. Will tweak later if rules differ.

File: assets/app/view/game/buy_trains.rb (hunk 18 above)
Comment (tobymao) [reply assets/app/view/game/buy_trains.rb:103]: you need to add skip true

**Comment by tobymao** [comment]
can you rebase?

File: models/user.rb (hunk 16 above)
Comment (crericha) [reply models/user.rb]: Fixed.

**Pull request merged**